Content/Items/Misc/Weapons.Soilgun.cs

```diff
@@ -0,0 +1,1488 @@
+            Projectile.ignoreWater = true;
+        }
+
+        public override void AI()
+        {
+            CurrentCharge++;
+
+            Vector2 armPos = owner.RotatedRelativePoint(owner.MountedCenter, true);
+
+            Vector2 barrelPos = armPos + Projectile.velocity * Projectile.width * 0.5f;
+
+            if (MaxCharge == 0f)
+            {
+                MaxCharge = owner.HeldItem.useAnimation;
+            }
+
+            if (!CanHold)
+            {
+                if (CurrentCharge >= MaxCharge)
+                    ShootThings(barrelPos);
+
+                Projectile.Kill();
+            }
+            if (CurrentCharge == MaxCharge)
+            {
+                //maybe better sound here
+                SoundEngine.PlaySound(SoundID.MaxMana, Projectile.position);
+                CombatText.NewText(owner.getRect(), new Color(151, 107, 75), "Loaded!", true, true);
+                for (int i = 0; i < 9; i++)
+                {
+                    Dust.NewDust(barrelPos, 4, 8, DustID.Dirt, 0f, 0f, default, default, Main.rand.NextFloat(0.9f, 1.2f));
+                }
+            }
+
+            owner.ChangeDir(Projectile.direction);
+            owner.heldProj = Projectile.whoAmI;
+            owner.itemTime = 2;
+            owner.itemAnimation = 2;
+            owner.itemRotation = Utils.ToRotation(Projectile.velocity * Projectile.direction);
+            Projectile.timeLeft = 2;
+            Projectile.position = armPos - Projectile.Size * 0.5f;
+            Projectile.rotation = Utils.ToRotation(Projectile.velocity);
+
+            if (Projectile.spriteDirection == -1)
+                Projectile.rotation += 3.1415927f;
+
+            Projectile.spriteDirection = Projectile.direction;
+
+            if (Main.myPlayer == Projectile.owner)
+            {
+                float interpolant = Utils.GetLerpValue(5f, 25f, Projectile.Distance(Main.MouseWorld), true);
+                Vector2 oldVelocity = Projectile.velocity;
+                Projectile.velocity = Vector2.Lerp(Projectile.velocity, owner.DirectionTo(Main.MouseWorld), interpolant);
+                if (Projectile.velocity != oldVelocity)
+                {
+                    Projectile.netSpam = 0;
+                    Projectile.netUpdate = true;
+                }
+            }
+            if (CurrentCharge < 2)
+                return;
+            int DustFrequency = (int)(15 - (Utils.Clamp(CurrentCharge / 5, 0, 12)));
+            if (Main.rand.NextBool(Utils.Clamp(DustFrequency, 1, 15)))
+            {
+                Dust dust = Dust.NewDustDirect(barrelPos, 2, 8, DustID.Dirt, 0f, 0f);
+                dust.scale = Main.rand.NextFloat(0.8f, 1.2f);
+                dust.noGravity = false;
+                if (Main.rand.NextBool(5))
+                    Dust.NewDustDirect(barrelPos, 2, 8, ModContent.DustType<Dusts.Sand>(), 0, 0, 125, default, 0.5f);
+            }
+        }
+
+        public void ShootThings(Vector2 position)
```

Comment:
This method name is unacceptable. Please use a more descriptive name. In general, avoid using words such as "thing, stuff, etc." in names.